fah_xchem/analysis/diffnet.py

```diff
@@ -0,0 +1,311 @@
+import logging
+from math import log
+from typing import List, Optional
+
+import networkx as nx
+import numpy as np
+
+from ..schema import (
+    Compound,
+    CompoundAnalysis,
+    CompoundMicrostate,
+    Microstate,
+    MicrostateAnalysis,
+    PointEstimate,
+    TransformationAnalysis,
+)
+from .exceptions import AnalysisError, ConfigurationError, InsufficientDataError
+
+
+def pIC50_to_DG(pIC50: float, s_conc: float = 375e-9, Km: float = 40e-6) -> float:
+    """
+    Converts IC50 (in M units) to DG
+    Parameters
+    ----------
+    pIC50 : float
+        pIC50
+    s_conc : float, default=375E-9
+        Substrate concentration in M
+    Km : float, default=40E-6
+        Substrate concentration for half-maximal enzyme activity
+    Returns
+    -------
+    type
+        Description of returned object.
+    """
+    ic50 = 10 ** -pIC50
+
+    if ic50 > 1e-5:
+        logging.warning("Expecting IC50 in M units. Please check.")
+
+    Ki = ic50 / (1 + s_conc / Km)
+    return log(Ki)
+
+
+def get_compound_free_energy(microstates: List[MicrostateAnalysis]) -> PointEstimate:
+    r"""
+    Compute compound free energy from microstate free energies and
+    microstate free energy penalties.
+
+    The sum over microstates is given by
+
+    .. math:: g_c = - \log \sum_{i} \exp \left[-(s_{ci} + g_{ci}) \right]
+
+    Parameters
+    ----------
+    microstates : list of MicrostateAnalysis
+        Microstate free energies
+    """
+    penalized_free_energies = [
+        microstate.free_energy + microstate.microstate.free_energy_penalty
```

Comment:
`microstate.microstate`?

Reply:
Yep, I don't love this either, but there is some nesting here because the first `microstate` is `MicrostateAnalysis`, which has the metadata and analysis results bundled together:
```python
class MicrostateAnalysis(Model):
    microstate: Microstate
    free_energy: Optional[PointEstimate]
```
This is to avoid needing to repeat the fields in `Microstate` inside of `MicrostateAnalysis`. We could rename the nested field, e.g. if `microstate.info` is easier on the eyes/fingers.